core/storage/suggestion/gae_models.py

```diff
@@ -0,0 +1,267 @@
+    # contains the data necessary to add the new entity.
+    payload = ndb.JsonProperty(required=True, indexed=False)
+
+    @classmethod
+    def get_instance_id(
+            cls, entity_type, thread_id, entity_id):
+        """Concatenates various parameters and gives the ID of the suggestion
+        model.
+
+        Args:
+            entity_type: str. The type of entity being edited/added.
+            thread_id: str. The ID of the feedback thread linked to the
+                suggestion.
+            entity_id: str. The ID of the entity being edited.
+
+        Returns:
+            str. The full instance ID for the suggestion.
+        """
+        return SUGGESTION_ID_DELIMITER.join([entity_type, thread_id, entity_id])
+
+    @classmethod
+    def create(
+            cls, suggestion_type, entity_type, status,
+            suggestion_params, author_id, reviewer_id, thread_id,
+            assigned_reviewer_id, payload):
+        """Creates a new SuggestionModel entry.
+
+        Args:
+            suggestion_type: str. The type of the suggestion.
+            entity_type: str. The type of entity being edited/added.
+            status: str. The status of the suggestion.
+            suggestion_params: dict. Additional parameters for the suggestion
+            author_id: str. The ID of the user who submitted the suggestion.
+            reviewer_id: str. The ID of the reviewer who has accepted the
+                suggestion.
+            thread_id: str. The ID of the feedback thread linked to the
+                suggestion.
+            assigned_reviewer_id: str. The ID of the user assigned to
+                review the suggestion.
+            payload: dict. The actual content of the suggestion.
+
+        Raises:
+            Exception: There is already a suggestion with the given id.
+        """
+        instance_id = cls.get_instance_id(
+            entity_type, thread_id, suggestion_params['entity_id'])
+        suggestion_params = SuggestionParams(
+            contribution_type=suggestion_params['contribution_type'],
+            contribution_domain=suggestion_params['contribution_domain'],
+            entity_id=suggestion_params['entity_id'],
+            entity_version=suggestion_params['entity_version'])
+
+        if cls.get_by_id(instance_id):
+            raise Exception('There is already a suggestion with the given'
+                            ' id: %s' % instance_id)
+
+        cls(id=instance_id, suggestion_type=suggestion_type,
+            entity_type=entity_type, status=status,
+            suggestion_params=suggestion_params, author_id=author_id,
+            reviewer_id=reviewer_id, thread_id=thread_id,
+            assigned_reviewer_id=assigned_reviewer_id, payload=payload).put()
+
+    @classmethod
+    def get_suggestions_by_type(cls, suggestion_type):
+        """Gets all suggestions of a particular type
+
+        Args:
+            suggestion_type: str. The type of the suggestions.
```

Comment:
I just realized I'm confused about something. What is the difference between suggestion type and contribution type? Is contribution type used only for scoring? If so then it needs to be renamed to reflect that, I think (score_category perhaps).

Reply:
Discussed. Contribution_type is used only for scoring. Renamed it to score_category.